#!/usr/bin/Rscript

################ WGCNA RNA-seq Analysis project snapper ############
####################################################
ls()
rm(list=ls())
ls()
#source("http://bioconductor.org/biocLite.R") 
#biocLite(c("AnnotationDbi", "impute", "GO.db", "preprocessCore")) 
#install.packages("WGCNA")

#install.packages("scales")
#install.packages("assertthat")
library("assertthat")
library("scales")
library("WGCNA")
setwd("./")
# The following setting is important, do not omit.
options(stringsAsFactors = FALSE);

#Read in the expr data set
logcpm_test<-read.table("02_data/logcpm_edger_snapper.txt", header=T)
# Take a quick look at what is in the data set:
dim(logcpm_test)

#=====================================================================================
#
#  Code chunk 2
#
#=====================================================================================
datExpr0 = as.data.frame(t(logcpm_test))
head(datExpr0)
#=====================================================================================
#
#  Code chunk 3
#
#=====================================================================================


gsg = goodSamplesGenes(datExpr0, verbose = 3);
gsg$allOK

#=====================================================================================
#
#  Code chunk 4
#
#=====================================================================================


if (!gsg$allOK)
{
  # Optionally, print the gene and sample names that were removed:
  if (sum(!gsg$goodGenes)>0) 
    printFlush(paste("Removing genes:", paste(names(datExpr0)[!gsg$goodGenes], collapse = ", ")));
  if (sum(!gsg$goodSamples)>0) 
    printFlush(paste("Removing samples:", paste(rownames(datExpr0)[!gsg$goodSamples], collapse = ", ")));
  # Remove the offending genes and samples from the data:
  datExpr0 = datExpr0[gsg$goodSamples, gsg$goodGenes]
}

#=====================================================================================
#
#  Code chunk 5
#
#=====================================================================================


sampleTree = hclust(dist(datExpr0), method = "average");
# Plot the sample tree: Open a graphic output window of size 12 by 9 inches
# The user should change the dimensions if the window is too large or too small.
sizeGrWindow(12,9)
#pdf(file = "Plots/sampleClustering.pdf", width = 12, height = 9);
par(cex = 0.6);
par(mar = c(0,4,2,0))
plot(sampleTree, main = "Sample clustering to detect outliers", sub="", xlab="", cex.lab = 1.5, 
     cex.axis = 1.5, cex.main = 2)

#=====================================================================================
#
#  Code chunk 6: check outliers
#
#=====================================================================================


# Plot a line to show the cut
abline(h = 250, col = "red");
# Determine cluster under the line
clust = cutreeStatic(sampleTree, cutHeight = 250, minSize = 10)
table(clust)
# clust 1 contains the samples we want to keep.
keepSamples = (clust==1)
datExpr = datExpr0[keepSamples, ]
nGenes = ncol(datExpr)
nSamples = nrow(datExpr)


variancedatExpr=as.vector(apply(as.matrix(datExpr),2,var, na.rm=T))
no.presentdatExpr=as.vector(apply(!is.na(as.matrix(datExpr)),2, sum) )
# Another way of summarizing the number of pressent entries
table(no.presentdatExpr)

# Keep only genes whose variance is non-zero and have at least 4 present entries
KeepGenes= variancedatExpr>0.1 & no.presentdatExpr>=4
table(KeepGenes)
datExpr=datExpr[, KeepGenes]
View(datExpr)
names(datExpr)
name_datExpr <-colnames(datExpr)
head(name_datExpr)
write.table(name_datExpr,file="03_results/genes_used_wgcna.txt")
#=====================================================================================
#
#  Code chunk 7: add traits
#
#=====================================================================================


allTraits = read.csv("02_data/traitmusclesubset_DC.csv",sep=";", header=T,dec=",");
names(allTraits)
str(allTraits)
# Form a data frame analogous to expression data that will hold the clinical traits.

femaleSamples = rownames(datExpr);
femaleSamples
allTraits$Ind
traitRows = match(femaleSamples, allTraits$Ind);

datTraits = allTraits[traitRows, -1];
rownames(datTraits) = allTraits[traitRows, 1];
str(datTraits)

collectGarbage();


#=====================================================================================
#
#  Code chunk 8
#
#=====================================================================================


# Re-cluster samples
sampleTree2 = hclust(dist(datExpr), method = "average")
# Convert traits to a color representation: white means low, red means high, grey means missing entry
traitColors = numbers2colors(datTraits,signed= FALSE);
# Plot the sample dendrogram and the colors underneath.
pdf("03_results/wgcna/figures/dendo_heatmap_subset.pdf",width=12,height=9)
par(mar=c(1, 10, 1, 1))
plotDendroAndColors(sampleTree2, traitColors,
                    groupLabels = names(datTraits), 
                    main = "Sample dendrogram and trait heatmap")
?plotDendroAndColors()
dev.off()

#=====================================================================================
#
#  Code chunk 9
#
#=====================================================================================


save(datExpr, datTraits, file = "02_data/Snapper-dataInput_subset.Rda")



########################## Module construction step-by-step #################################
################################################################################
#=====================================================================================
#
#  Code chunk 1
#
#=====================================================================================

#setting is important, do not omit.
options(stringsAsFactors = FALSE);
# Allow multi-threading within WGCNA. At present this call is necessary.
# Any error here may be ignored but you may want to update WGCNA if you see one.
# Caution: skip this line if you run RStudio or other third-party R environments.
# See note above.
#enableWGCNAThreads()
# Load the data saved in the first part
lnames = load(file = "02_data/Snapper-dataInput_subset.Rda");
#The variable lnames contains the names of loaded variables.
lnames


#=====================================================================================
#
#  Code chunk 2
#
#=====================================================================================

load("02_data/Snapper-dataInput_subset.Rda")
# Choose a set of soft-thresholding powers
powers = c(c(1:10), seq(from = 10, to=24, by=2))
# Call the network topology analysis function
sft = pickSoftThreshold(datExpr, powerVector = powers, verbose = 5,networkType="signed")
# Plot the results:
sizeGrWindow(9, 5)
par(mfrow = c(1,2));
cex1 = 0.9;
# Scale-free topology fit index as a function of the soft-thresholding power
plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
     main = paste("Scale independence"));
text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     labels=powers,cex=cex1,col="red");
# this line corresponds to using an R^2 cut-off of h
abline(h=0.80,col="red")
# Mean connectivity as a function of the soft-thresholding power
plot(sft$fitIndices[,1], sft$fitIndices[,5],
     xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
     main = paste("Mean connectivity"))
text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")
save(sft,file="02_data/sft_signed_snapper_subset.Rda")
View(sft$fitIndices)
#=====================================================================================
#
#  Code chunk 3
#
#=====================================================================================


softPower = 24;
adjacency = adjacency(datExpr, power = softPower,type="signed");


#=====================================================================================
#
#  Code chunk 4
#
#=====================================================================================
load("sft_signed_snapper_subset.Rda")
load("Snapper-networkConstruction-stepByStep_subset.Rda")
# Turn adjacency into topological overlap

TOM = TOMsimilarity(adjacency,TOMType="signed");
dissTOM = 1-TOM

load("TOM.Rda") # did on datarmor


#=====================================================================================
#
#  Code chunk 5
#
#=====================================================================================


# Call the hierarchical clustering function
geneTree = hclust(as.dist(dissTOM), method = "average");
# Plot the resulting clustering tree (dendrogram)
sizeGrWindow(12,9)
plot(geneTree, xlab="", sub="", main = "Gene clustering on TOM-based dissimilarity",
     labels = FALSE, hang = 0.04);


#=====================================================================================
#
#  Code chunk 6
#
#=====================================================================================


# We like large modules, so we set the minimum module size relatively high:
minModuleSize = 30;
# Module identification using dynamic tree cut:
dynamicMods =cutreeDynamic(dendro = geneTree, distM = dissTOM,
                            deepSplit = 2, pamRespectsDendro = FALSE,
                            minClusterSize = minModuleSize);

?cutreeDynamic()
head(colorDynamicTOM)
table(dynamicMods)
?cutreeDynamic()


#=====================================================================================
#
#  Code chunk 7
#
#=====================================================================================


# Convert numeric lables into colors
dynamicColors = labels2colors(dynamicMods)
table(dynamicColors)
# Plot the dendrogram and colors underneath
sizeGrWindow(8,6)
plotDendroAndColors(geneTree, dynamicColors, "Dynamic Tree Cut",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
                    main = "Gene dendrogram and module colors")


#=====================================================================================
#
#  Code chunk 8
#
#=====================================================================================


# Calculate eigengenes
MEList = moduleEigengenes(datExpr, colors = dynamicColors)
MEs = MEList$eigengenes
# Calculate dissimilarity of module eigengenes
MEDiss = 1-cor(MEs);
# Cluster module eigengenes
METree = hclust(as.dist(MEDiss), method = "average");

# Plot the result
sizeGrWindow(7, 6)
plot(METree, main = "Clustering of module eigengenes",
     xlab = "", sub = "")


#=====================================================================================
#
#  Code chunk 9
#
#=====================================================================================


MEDissThres = 0.25
# Plot the cut line into the dendrogram
abline(h=MEDissThres, col = "red")
# Call an automatic merging function
merge = mergeCloseModules(datExpr, dynamicColors, cutHeight = MEDissThres, verbose = 3)
# The merged module colors
mergedColors = merge$colors;
# Eigengenes of the new merged modules:
mergedMEs = merge$newMEs;


#=====================================================================================
#
#  Code chunk 10
#
#=====================================================================================


sizeGrWindow(12, 9)
#pdf(file = "Plots/geneDendro-3.pdf", wi = 9, he = 6)
plotDendroAndColors(geneTree, cbind(dynamicColors, mergedColors),
                    c("Dynamic Tree Cut", "Merged dynamic"),
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)
#dev.off()


#=====================================================================================
#
#  Code chunk 11
#
#=====================================================================================


# Rename to moduleColors
moduleColors = mergedColors
# Construct numerical labels corresponding to the colors
colorOrder = c("grey", standardColors(50));
moduleLabels = match(moduleColors, colorOrder)-1;
MEs = mergedMEs;
# Save module colors and labels for use in subsequent parts
save(MEs, moduleLabels, moduleColors, geneTree, file = "02_data/Snapper-networkConstruction-stepByStep_subset.Rda")


############### Relating genes with traits and identifying important traits ##############
##########################################################################################

#=====================================================================================
#
#  Code chunk 1
#
#=====================================================================================

# The following setting is important, do not omit.
options(stringsAsFactors = FALSE);
# Load the expression and trait data saved in the first part
lnames = load(file = "02_data/Snapper-dataInput_subset.Rda");
#The variable lnames contains the names of loaded variables.
lnames
# Load network data saved in the second part.
lnames = load(file = "02_data/Snapper-networkConstruction-stepByStep_subset.Rda");
lnames = load(file = "02_data/sft_signed_snapper_subset.Rda")
lnames

str(MEs)
#=====================================================================================
#
#  Code chunk 2
#
#=====================================================================================


# Define numbers of genes and samples
nGenes = ncol(datExpr);
nSamples = nrow(datExpr);
# Recalculate MEs with color labels
MEs0 = moduleEigengenes(datExpr, moduleColors)$eigengenes
MEs = orderMEs(MEs0)
moduleTraitCor = cor(MEs, datTraits, use = "p");
moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples);
#moduleTraitPvalue[moduleTraitPvalue>0.049]<-0
#moduleTraitCor[moduleTraitCor > -0.6 & moduleTraitCor < 0.6]<-0
head(moduleTraitCor)
head(moduleTraitPvalue)
#=====================================================================================
#
#  Code chunk 3
#
#=====================================================================================


sizeGrWindow(25,15)


# Will display correlations and their p-values
#moduleTraitPvalue[moduleTraitPvalue >0.049]=NA #test
textMatrix =  paste(signif(moduleTraitCor, 2), "\n(",
                    signif(moduleTraitPvalue, 1), ")", sep = "");



dim(textMatrix) = dim(moduleTraitCor)
par(mar = c(6, 10, 3, 3));
# Display the correlation values within a heatmap plot
labeledHeatmap(Matrix = moduleTraitCor,
               xLabels = names(datTraits),
               yLabels = names(MEs),
               ySymbols = names(MEs),
               colorLabels = FALSE,
               colors = blueWhiteRed(50),
               textAdj = c(0.5, 0.5),
               textMatrix = textMatrix,
               setStdMargins = FALSE,
               cex.text = 0.5,
               zlim = c(-1,1),
               main = paste("Module-trait relationships"))




#install.packages("corrplot")
library(corrplot)
par(mar = c(1,1, 1 ,1));

moduleTraitCor[moduleTraitCor > -0.6 & moduleTraitCor < 0.6]<-0
df_module<-as.data.frame(moduleTraitCor)

#write.table(df_module,file="test_module.txt", quote=F)
new_mat <- as.data.frame(moduleTraitCor[ rownames(moduleTraitCor) %in% c("MEmagenta4",
                                                                        "MEmediumpurple4",
                                                                        "MEdarkorange2",
                                                                        "MEsalmon4",
                                                                        "MEbrown4",
                                                                        "MEblue",
                                                                        "MEdarkviolet",
                                                                        "MEmediumorchid",
                                                                        "MEbisque4",
                                                                        "MEdarkseagreen3",
                                                                        "MElavenderblush2",
                                                                        "MEantiquewhite2",
                                                                        "MEcoral"),])
new_mat$Weight <-NULL
new_mat$HSI <-NULL
new_mat$Gluc<-NULL
new_mat<-new_mat[c("Temperature","Genotype","LGR","SGR","MCHC","Hb")]
row.names(new_mat)<-gsub("ME", "", row.names(new_mat)) 

View(new_mat)
mat<-as.matrix(new_mat)
tiff(file = "correlation_matrix.tiff", width = 20, height = 20, units = "cm", res = 300)
corrplot(mat, method = "number",tl.col = "black",cl.ratio = 0.25, cl.align = "r")
dev.off()
#=====================================================================================
#
#  Code chunk 4
#
#=====================================================================================


# Define variable weight containing the weight column of datTrait
temperature = as.data.frame(datTraits$Temperature);
names(temperature) = "temperature"
str(temperature)
# names (colors) of the modules
modNames = substring(names(MEs), 3)
str(modNames)
geneModuleMembership = as.data.frame(cor(datExpr, MEs, use = "p"));
str(geneModuleMembership)


MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nSamples));
str(MMPvalue)
names(geneModuleMembership) = paste("MM", modNames, sep="");
names(MMPvalue) = paste("p.MM", modNames, sep="");

geneTraitSignificance = as.data.frame(cor(datExpr, temperature, use = "p"));
GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), nSamples));

names(geneTraitSignificance) = paste("GS.", names(temperature), sep="");
names(GSPvalue) = paste("p.GS.", names(temperature), sep="");


# plot module significance ## to implement

#=====================================================================================
#
#  Code chunk 5
#
#=====================================================================================

dev.off()
module = "darkviolet"
column = match(module, modNames);
moduleGenes = moduleColors==module;

tiff(file = "module_darkviolet.tiff", width = 15, height = 15, units = "cm", res = 300)
plot<-verboseScatterplot(abs(geneModuleMembership[moduleGenes, column]),
                   abs(geneTraitSignificance[moduleGenes, 1]),
                   xlab = paste("Module Membership in", module, "module"),
                   ylab = "Gene significance for temperature",
                   main = paste(""),
                   cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2, col = module)
plot + text(0.4, 0.85, "cor = 0.95\np < 0.001",cex = 1.5,font=2)
dev.off()


module = "bisque4"
column = match(module, modNames);
moduleGenes = moduleColors==module;
tiff(file = "module_bisque4.tiff", width = 15, height = 15, units = "cm", res = 300)
#sizeGrWindow(7, 7);
#par(mfrow = c(1,1));
plot<-verboseScatterplot(abs(geneModuleMembership[moduleGenes, column]),
                   abs(geneTraitSignificance[moduleGenes, 1]),
                   xlab = paste("Module Membership in", module, "module"),
                   ylab = "Gene significance for temperature",
                   main = NULL,
                   sub = NULL,
                   cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2, col = module)
plot + text(0.2, 0.85, "cor = 0.91\np < 0.001",cex = 1.5,font=2)
dev.off()



annot = read.csv(file = "01_info_files/snapper.annotation_copie.csv",sep=";",header=T);
dim(annot)
names(annot)
probes = names(datExpr)
probes
probes2annot = match(probes, annot$name)
# The following is the number or probes without annotation:
sum(is.na(probes2annot))
# Should return 0.



#=====================================================================================
#
#  Code chunk 9
#
#=====================================================================================


# Create the starting data frame
geneInfo0 = data.frame(geneID = probes,
                       geneSymbol = annot$Sp[probes2annot],
                       LocusLinkID = annot$ID[probes2annot],
                       moduleColor = moduleColors,
                       geneTraitSignificance,
                       GSPvalue)
# Order modules by their significance for temperature
modOrder = order(-abs(cor(MEs, genotype, use = "p")));

# get relevant modules
# Get the corresponding Locuis Link IDs
allLLIDs = annot$name[probes2annot];
# $ Choose interesting modules
intModules = c("magenta4", "mediumpurple4", "darkorange2",
               "salmon4",
               "brown4",
               "blue",
               "darkviolet",
               "mediumorchid",
               "bisque4",
               "darkseagreen3",
               "lavenderblush2",
               "coral",
               "antiquewhite2")
for (module in intModules)
{
  # Select module probes
  modGenes = (moduleColors==module)
  # Get their entrez ID codes
  modLLIDs = allLLIDs[modGenes];
  # Write them into a file
  fileName = paste("module_", module, ".txt", sep="");
  write.table(as.data.frame(modLLIDs), file = fileName,
              row.names = FALSE, col.names = FALSE,quote=FALSE)
}

# A
# Add module membership information in the chosen order
for (mod in 1:ncol(geneModuleMembership))
{
  oldNames = names(geneInfo0)
  geneInfo0 = data.frame(geneInfo0, geneModuleMembership[, modOrder[mod]], 
                         MMPvalue[, modOrder[mod]]);
  names(geneInfo0) = c(oldNames, paste("MM.", modNames[modOrder[mod]], sep=""),
                       paste("p.MM.", modNames[modOrder[mod]], sep=""))
}

# Order the genes in the geneInfo variable first by module color, then by geneTraitSignificance
geneOrder = order(geneInfo0$moduleColor, -abs(geneInfo0$GS.temperature));
geneInfo = geneInfo0[geneOrder, ]
write.csv(geneInfo, file = "geneinfo_snapper_temperature.csv")
#=====================================================================================
#
#  Code chunk 5
#
#=====================================================================================
######## same for genotype

# Define variable weight containing the weight column of datTrait
genotype = as.data.frame(datTraits$Genotype);
names(genotype) = "genotype"
str(genotype)
# names (colors) of the modules
modNames = substring(names(MEs), 3)

geneModuleMembership = as.data.frame(cor(datExpr, MEs, use = "p"));
str(geneModuleMembership)


MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nSamples));
str(MMPvalue)
names(geneModuleMembership) = paste("MM", modNames, sep="");
names(MMPvalue) = paste("p.MM", modNames, sep="");

geneTraitSignificance = as.data.frame(cor(datExpr, genotype, use = "p"));
GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), nSamples));

names(geneTraitSignificance) = paste("GS.", names(genotype), sep="");
names(GSPvalue) = paste("p.GS.", names(genotype), sep="");

#### same for genotype
module = "darkorange2"
column = match(module, modNames);
moduleGenes = moduleColors==module;

tiff(file = "module_darkorgange2.tiff", width = 15, height = 15, units = "cm", res = 300)
plot<-verboseScatterplot(abs(geneModuleMembership[moduleGenes, column]),
                   abs(geneTraitSignificance[moduleGenes, 1]),
                   xlab = paste("Module Membership in", module, "module"),
                   ylab = "Gene significance for genotype",
                   main = paste(""),
                   cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2, col = module)

plot + text(0.55, 0.85, "cor = 0.67\np < 0.001",cex = 1.5,font=2)
dev.off()

module = "antiquewhite2"
column = match(module, modNames);
moduleGenes = moduleColors==module;

tiff(file = "module_antiquewhite2.tiff", width = 15, height = 15, units = "cm", res = 300)
plot<-verboseScatterplot(abs(geneModuleMembership[moduleGenes, column]),
                   abs(geneTraitSignificance[moduleGenes, 1]),
                   xlab = paste("Module Membership in", module, "module"),
                   ylab = "Gene significance for genotype",
                   main = paste(""),
                   cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2, col = module)
plot + text(0.3, 0.8, "cor = 0.49\np < 0.001",cex = 1.5,font=2)
dev.off()
#=====================================================================================
#
#  Code chunk 6
#
#=====================================================================================


names(datExpr)


#=====================================================================================
#
#  Code chunk 7
#
#=====================================================================================



names(datExpr)[moduleColors=="blue"]



#=====================================================================================
#
#  Code chunk 8
#
#=====================================================================================


annot = read.csv(file = "01_info_files/snapper.annotation_copie.csv",sep=";",header=T);
dim(annot)
names(annot)
probes = names(datExpr)
probes
probes2annot = match(probes, annot$name)
# The following is the number or probes without annotation:
sum(is.na(probes2annot))
# Should return 0.



#=====================================================================================
#
#  Code chunk 9
#
#=====================================================================================


# Create the starting data frame
geneInfo0 = data.frame(geneID = probes,
                       geneSymbol = annot$Sp[probes2annot],
                       LocusLinkID = annot$ID[probes2annot],
                       moduleColor = moduleColors,
                       geneTraitSignificance,
                       GSPvalue)
# Order modules by their significance for temperature
modOrder = order(-abs(cor(MEs, genotype, use = "p")));

# get relevant modules
# Get the corresponding Locuis Link IDs
allLLIDs = annot$name[probes2annot];
# $ Choose interesting modules
intModules = c("magenta4", "mediumpurple4", "darkorange2",
               "salmon4",
               "brown4",
               "blue",
               "darkviolet",
               "mediumorchid",
               "bisque4",
               "darkseagreen3",
                "lavenderblush2",
               "coral",
                "antiquewhite2")
for (module in intModules)
{
  # Select module probes
  modGenes = (moduleColors==module)
  # Get their entrez ID codes
  modLLIDs = allLLIDs[modGenes];
  # Write them into a file
  fileName = paste("module_", module, ".txt", sep="");
  write.table(as.data.frame(modLLIDs), file = fileName,
              row.names = FALSE, col.names = FALSE,quote=FALSE)
}

# A
# Add module membership information in the chosen order
for (mod in 1:ncol(geneModuleMembership))
{
  oldNames = names(geneInfo0)
  geneInfo0 = data.frame(geneInfo0, geneModuleMembership[, modOrder[mod]], 
                         MMPvalue[, modOrder[mod]]);
  names(geneInfo0) = c(oldNames, paste("MM.", modNames[modOrder[mod]], sep=""),
                       paste("p.MM.", modNames[modOrder[mod]], sep=""))
}

# Order the genes in the geneInfo variable first by module color, then by geneTraitSignificance
geneOrder = order(geneInfo0$moduleColor, -abs(geneInfo0$GS.genotype));
geneInfo = geneInfo0[geneOrder, ]

# Test visualize cluster
#par(mar = c(3, 3, 3, 3));
#distance=1-(1+cor(MEs,use="p"))/2
#MDS = cmdscale(as.dist(distance),2)
#cluster = hclust(as.dist(distance),method="a")
#plot(cluster)
#plot(MDS,col=moduleColors,pch=19,cex=2, xlab="MDS1", ylab="MDS2")
#=====================================================================================
#
#  Code chunk 10
#
#=====================================================================================


write.csv(geneInfo, file = "geneinfo_snapper_genotype.csv")




########################## Visualize network ##########################
#######################################################################

#=====================================================================================
#
#  Code chunk 1
#
#=====================================================================================

# Load the expression and trait data saved in the first part
lnames = load(file = "02_data/Snapper-dataInput_subset.Rda");
#The variable lnames contains the names of loaded variables.
lnames
# Load network data saved in the second part.
lnames = load(file = "02_data/Snapper-networkConstruction-stepByStep_subset.Rda");
lnames
nGenes = ncol(datExpr)
nSamples = nrow(datExpr)
# load SFT  ### not made in the proper package
lnames=load(file="02_data/sft_signed_snapper_subset.Rda")
lnames
#=====================================================================================
#
#  Code chunk 2
#
#=====================================================================================


# Calculate topological overlap anew: this could be done more efficiently by saving the TOM
# calculated during module detection, but let us do it again here.
#dissTOM = 1-TOMsimilarityFromExpr(datExpr, power = 22);
# Transform dissTOM with a power to make moderately strong connections more visible in the heatmap
plotTOM = dissTOM^7;
# Set diagonal to NA for a nicer plot
diag(plotTOM) = NA;
# Call the plot function
sizeGrWindow(9,9)
TOMplot(plotTOM, geneTree, moduleColors, main = "Network heatmap plot, all genes")


#=====================================================================================
#
#  Code chunk 3
#
#=====================================================================================


nSelect = 400
# For reproducibility, we set the random seed
set.seed(10);
select = sample(nGenes, size = nSelect);
selectTOM = dissTOM[select, select];
# There's no simple way of restricting a clustering tree to a subset of genes, so we must re-cluster.
selectTree = hclust(as.dist(selectTOM), method = "average")
selectColors = moduleColors[select];
# Open a graphical window
sizeGrWindow(9,9)
# Taking the dissimilarity to a power, say 10, makes the plot more informative by effectively changing 
# the color palette; setting the diagonal to NA also improves the clarity of the plot
plotDiss = selectTOM^7;
diag(plotDiss) = NA;
TOMplot(plotDiss, selectTree, selectColors, main = "Network heatmap plot, selected genes")


#=====================================================================================
#
#  Code chunk 4
#
#=====================================================================================


# Recalculate module eigengenes
MEs = moduleEigengenes(datExpr, moduleColors)$eigengenes
# Isolate weight from the clinical traits
weight = as.data.frame(datTraits$Temperature);
names(weight) = "weight"
# Add the weight to existing module eigengenes
MET = orderMEs(cbind(MEs, weight))
# Plot the relationships among the eigengenes and the trait
sizeGrWindow(5,7.5);
par(cex = 0.9)
plotEigengeneNetworks(MET, "", marDendro = c(0,4,1,2), marHeatmap = c(3,4,1,2), cex.lab = 0.8, xLabelsAngle
                      = 90)


#=====================================================================================
#
#  Code chunk 5
#
#=====================================================================================


# Plot the dendrogram
sizeGrWindow(6,6);
par(cex = 1.0)
plotEigengeneNetworks(MET, "Eigengene dendrogram", marDendro = c(0,4,2,0),
                      plotHeatmaps = FALSE)
# Plot the heatmap matrix (note: this plot will overwrite the dendrogram plot)
par(cex = 1.0)
plotEigengeneNetworks(MET, "Eigengene adjacency heatmap", marHeatmap = c(3,4,2,2),
                      plotDendrograms = FALSE, xLabelsAngle = 90)



############### Export  ######################################
##########################################################################################


### selection best genes ? ######
#################################
# test
#modular membership
#essaikME=as.data.frame(signedKME(datExpr,MEs,outputColumnName = "kME", corFnc = "cor", corOptions = "use = 'p'"))


##### Export to Visant #######################################################################
###############################################################################################

# Load the expression and trait data saved in the first part
lnames = load(file = "02_data/Snapper-dataInput_subset.Rda");
#The variable lnames contains the names of loaded variables.
lnames
# Load network data saved in the second part.
lnames = load(file = "02_data/Snapper-networkConstruction-stepByStep_subset.Rda");
lnames
load("02_data/TOM.Rda")
#TOM = TOMsimilarityFromExpr(datExpr, power = 22); ###a verifier
annot = read.csv(file = "01_info_files/snapper.annotation.csv",sep=";",header=T);

# Select module
module = "darkolivegreen";

# Select module probes
probes = names(datExpr)
inModule = (moduleColors==module);
modProbes = probes[inModule];
modTOM = TOM[inModule, inModule];
dimnames(modTOM) = list(modProbes, modProbes)


nTop = 30;
IMConn = softConnectivity(datExpr[, modProbes]);
top = (rank(-IMConn) <= nTop)
vis = exportNetworkToVisANT(modTOM[top, top],
                            file = paste("VisANTInput-", module, "-top30.txt", sep=""),
                            weighted = TRUE,
                            threshold = 0,
                            probeToGene = data.frame(annot$geneID, annot$geneSymbolCor))


